### ISSUE TITLE
[BUG] Service bus client never recovers from Exception

### ISSUE THREAD

**Issue by michaelegner** [Azure/azure-sdk-for-java#21199]
**Describe the bug**
Service bus client stopped receiving / sending messages after `com.azure.core.amqp.implementation.handler.CustomIOHandler` logged `Exception occurred when handling event in super`. The exception was cought by class `CustomIOHandler` and not re-thrown. Only after restarting the application the receiving and sending of messages worked fine again.

***Exception or Stack Trace***
Example of the exception log
`2021-04-27T14:07:48.209 ERROR [vis.ops.service.order,,,] [,] 1 --- [ctor-executor-9] c.a.c.a.i.handler.CustomIOHandler : Exception occurred when handling event in super.`

Follow up errors in the log:
```
2021-04-27T14:12:20.996 ERROR [vis.ops.service.order,3a9c177ff84b05c2,cd4919acbbc37c52,true] [a6bfa2b7-ddf3-48db-9d6f-506f538df274,en-US] 1 --- [     parallel-1] c.a.core.amqp.implementation.RetryUtil   : ReactorConnection: Retries exhausted waiting for ACTIVE endpoint state on CBS node.
Retries exhausted: 3/3
2021-04-27T14:12:21.000  WARN [vis.ops.service.order,3a9c177ff84b05c2,cd4919acbbc37c52,true] [a6bfa2b7-ddf3-48db-9d6f-506f538df274,en-US] 1 --- [     parallel-1] c.z.v.o.s.o.d.AbstractMessageProducer    : Error signal: com.azure.messaging.servicebus.ServiceBusException: Retries exhausted: 3/3
```

**To Reproduce**
Error occurs randomly 

***Code Snippet***
n.a.

**Expected behavior**
Error should be logged in a way to ease the troubleshooting and the client should recover from it without requiring an application restart.

**Screenshots**
n.a.

**Setup (please complete the following information):**
- OS: Windows, Linux
- IDE : IntelliJ
- Java 11, Spring Boot 2.3
- azure-messaging-servicebus: 7.0.2

**Additional context**
See for similar issue https://github.com/Azure/azure-sdk-for-java/issues/18219

**Information Checklist**
Kindly make sure that you have added all the following information above and checkoff the required fields otherwise we will treat the issuer as an incomplete report
- [x] Bug Description Added
- [x] Repro Steps Added
- [x] Setup information Added

**Comment by joshfree** [comment]
@hemanttanwar could you please take a look?

**Comment by YijunXieMS** [comment]
Hi @michaelegner today we released azure-message-servicebus 7.2.2, which has fixes on connection retry problems. Could you try it?

**Comment by michaelegner** [comment]
@YijunXieMS in version 7.2.2 is fixed, but after upgrading to version 7.3.0 we are facing a similar issue.

After these log entries the connection to servicebus is lost and no messages are processed or sent to queue:
`2021-07-13T10:32:50.298 ERROR [vis.ops.service.consumer,,,] [,] 1 --- [tor-executor-19] c.a.c.a.i.RequestResponseChannel         : connectionId[MF_bac10b_1626110919047] linkName[cbs] Error in SendLinkHandler. Disposing unconfirmed sends.
The connection was inactive for more than the allowed 300000 milliseconds and is closed by container 'LinkTracker'. TrackingId:520f1ef9f82c42f9a92856967c260d1e_G31, SystemTracker:gateway7, Timestamp:2021-07-13T10:32:50, errorContext[NAMESPACE: czag01r89sb002-test.servicebus.windows.net. ERROR CONTEXT: N/A, PATH: $cbs, REFERENCE_ID: cbs:sender, LINK_CREDIT: 99]
2021-07-13T10:32:50.298  WARN [vis.ops.service.consumer,,,] [,] 1 --- [tor-executor-19] c.a.c.a.i.RequestResponseChannel:$cbs    : Retry #1. Transient error occurred. Retrying after 4511 ms.
The connection was inactive for more than the allowed 300000 milliseconds and is closed by container 'LinkTracker'. TrackingId:520f1ef9f82c42f9a92856967c260d1e_G31, SystemTracker:gateway7, Timestamp:2021-07-13T10:32:50, errorContext[NAMESPACE: czag01r89sb002-test.servicebus.windows.net. ERROR CONTEXT: N/A, PATH: $cbs, REFERENCE_ID: cbs:sender, LINK_CREDIT: 99]
2021-07-13T10:32:54.810  INFO [vis.ops.service.consumer,,,] [,] 1 --- [     parallel-1] c.a.c.a.i.RequestResponseChannel:$cbs    : Retry #1. Requesting from upstream.
2021-07-13T10:32:54.810  INFO [vis.ops.service.consumer,,,] [,] 1 --- [     parallel-1] c.a.c.a.i.RequestResponseChannel:$cbs    : namespace[MF_bac10b_1626110919047] entityPath[$cbs]: Connection not requested, yet. Requesting one.
2021-07-13T10:32:54.811  INFO [vis.ops.service.consumer,,,] [,] 1 --- [     parallel-1] c.a.c.a.i.RequestResponseChannel:$cbs    : namespace[MF_bac10b_1626110919047] entityPath[$cbs]: Setting next AMQP channel.
2021-07-13T10:32:54.811  INFO [vis.ops.service.consumer,,,] [,] 1 --- [     parallel-1] c.a.c.a.i.RequestResponseChannel:$cbs    : namespace[MF_bac10b_1626110919047] entityPath[$cbs]: Next AMQP channel received, updating 2 current subscribers
`

**Comment by Dav1dde** [comment]
I am seeing the same issue after trying to move a message into the dead letter queue. The same error happens and the client is unusable after this.

**Comment by gondar** [comment]
We just got hit by this issue and it flooded our Application Insights log adding extra 60 GB in 2 hours from single server. 

@hemanttanwar I saw your request for reproducing it in the referenced ticket. I am happy to assist you, however the challenge of this issue is that it seemed to happen quite randomly and rarely so static analysis of the issue would be helpful. Please contact us by email and we will also provide you with JFR recording from the time of the incident.

**Comment by ramya-rao-a** [comment]
@gondar Please contact Azure support with the logs and details that you have

**Comment by gondar** [comment]
Sure I will. Do I just reference this github issue or is there other
preferred way to ensure those logs and details reach you? Or maybe I should
reference ticket number here once I obtain it from Azure support?

m.

**Comment by ramya-rao-a** [comment]
You can reference this GitHub issue in the support ticket

**Comment by ramya-rao-a** [comment]
Hey @michaelegner 

It has about 10 months since this issue was reported. In this time, we have shipped several updates to the azure-messaging-servicebus package that included multiple performance and reliability improvements. Please upgrade to the latest version 7.6.0 and if you still experience this issue, please drop a comment here so that we can re-open the investigation.

@Dav1dde Your scenario might be slightly different from the one originally reported. Please try the latest version and if you still see problems, please open a new issue with code snippets that show your usage of the client and error details.

@gondar, Hopefully you got a chance to open the support ticket.

Thanks for your patience and thanks for reporting everyone.

**Comment by gondar** [comment]
Hey,
Yes ticket is opened. Ticket TrackingID#2201040050000678

In the ticket we provided link to the line in code that was failing for our incident at the begging of January (2 months ago):
https://github.com/Azure/azure-sdk-for-java/blob/3f31d68eed6fbe11516ca3afe3955c8840a6e974/sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/handler/CustomIOHandler.java#L39

This line and comment above seem to justify keeping ticket open. Clearly, when you resolve this race condition you can remove that log and this log is what caused a trouble for us by sending Gigabites of information to application insights.

**Comment by gondar** [comment]
Here some of the data provided in log ticket (where I gave both full logs and how to access it in appinsights).
```

message	customDimensions_ThreadName	customDimensions_LoggerName	count_	min_timestamp [UTC]	max_timestamp [UTC]	
Exception occurred when handling event in super.	reactor-executor-20760	com.azure.core.amqp.implementation.handler.CustomIOHandler	230474027	1/3/2022, 12:01:32.248 PM	1/3/2022, 5:06:09.223 PM	
connectionId[MF_fd25b2_1641211292498] linkName[cbs]: Unable to open send and receive link.	reactor-executor-20821	com.azure.core.amqp.implementation.RequestResponseChannel	1	1/3/2022, 1:01:46.601 PM	1/3/2022, 1:01:46.601 PM	
connectionId[MF_0f10eb_1641211292686] linkName[cbs]: Unable to open send and receive link.	reactor-executor-20826	com.azure.core.amqp.implementation.RequestResponseChannel	1	1/3/2022, 1:01:46.570 PM	1/3/2022, 1:01:46.570 PM	
connectionId[MF_22c749_1641211292670] linkName[cbs]: Unable to open send and receive link.	reactor-executor-20825	com.azure.core.amqp.implementation.RequestResponseChannel	1	1/3/2022, 1:01:43.898 PM	1/3/2022, 1:01:43.898 PM	
connectionId[MF_bfbe4d_1641211293873] linkName[cbs]: Unable to open send and receive link.	reactor-executor-20857	com.azure.core.amqp.implementation.RequestResponseChannel	1	1/3/2022, 1:01:43.710 PM	1/3/2022, 1:01:43.710 PM	
```

**Comment by ramya-rao-a** [comment]
Thank you @gondar 
We are in touch with the folks handling your support ticket and you will get updates there

**Comment by meredrica** [comment]
we effectively have the same problem. since we're sending and reading quite a lot of messages, once the reading connection goes down we fill our topic within a few hours and we have no way of detecting the error.

**Comment by gondar** [comment]
Yes. There is programatic way to reset connection. We have scheduled
process to restart connection every hour to mitigate the issue.
This is not gr8 though. The issue above caused huge number of logs which
feels was a bit different. I think there is another issue more directly
related to what you experience.

**Comment by kroger-ktomazin** [comment]
In addition to the errors cited above, we have also had issues when attempting to complete messages, we are getting the following errors:

Exception The lock supplied is invalid. Either the lock expired, or the message has already been removed from the queue. For more information please see https://aka.ms/ServiceBusExceptions . Reference:95da59b1-1fd7-4f0c-a3eb-fe8743447c5b, TrackingId:6d392167000000e50000003b621ba1d4_G1S1_B9S1, SystemTracker:sb-eus2-mcs-prod-integration:Queue:sbq-mk-out-payment-001, Timestamp:2022-03-01T00:44:57, errorContext[NAMESPACE: sb-eus2-mcs-prod-integration.servicebus.windows.net, PATH: sbq-mk-out-payment-001, REFERENCE_ID: sbq-mk-out-payment-001_c1aaaf_1645976823791, LINK_CREDIT: 0]"

I have not seen any of these errors since upgrading the one service to 7.5.2 so I'm assuming this has been fixed in that version or earlier.  Please confirm as we are about to deploy our remaining services with that version and it would be nice to know that we will no longer see this issue.
